File: packages/sdk/src/finance-admin.ts
/** Typed SDK for Domain 68 — Finance Admin Dashboard. */

export type FinanceRole = 'viewer' | 'operator' | 'finance_admin' | 'super_admin';
export type RefundStatus =
  | 'draft' | 'pending' | 'approved' | 'processing'
  | 'succeeded' | 'failed' | 'rejected' | 'reversed';
export type RefundCategory =
  | 'duplicate' | 'fraud' | 'dispute' | 'goodwill'
  | 'service_failure' | 'cancelled' | 'partial';
export type FinProvider = 'stripe' | 'paddle' | 'wallet' | 'manual';
export type HoldReason = 'kyc' | 'risk_review' | 'dispute' | 'fraud' | 'manual' | 'compliance';
export type HoldStatus = 'active' | 'released' | 'expired';

export interface FinRefund {
  id: string; reference: string; invoice_id?: string | null; payment_ref?: string | null;
  customer_id: string; amount_minor: number; currency: string;
  reason: string; category: RefundCategory; provider: FinProvider; status: RefundStatus;
  requested_by?: string | null; approved_by?: string | null;
  approved_at?: string | null; processed_at?: string | null; failure_reason?: string | null;
  meta: Record<string, unknown>; created_at: string; updated_at: string;
}
export interface FinHold {
  id: string; owner_id: string; amount_minor: number; currency: string;
  reason: HoldReason; status: HoldStatus; expires_at?: string | null;
  released_at?: string | null; released_by?: string | null; notes?: string | null;
  meta: Record<string, unknown>; created_at: string;
}
export interface FinControl {
  id: string; scope: 'global' | 'customer' | 'plan' | 'region';
  scope_key: string; control_key: string; value: Record<string, unknown>; enabled: boolean;
  updated_by?: string | null; updated_at: string;
}
export interface FinLedgerEntry {
  id: string; occurred_at: string; account: string; owner_id?: string | null;
  ref_kind: string; ref_id?: string | null; direction: 'credit' | 'debit';
  amount_minor: number; currency: string; description?: string | null;
  actor_id?: string | null; meta: Record<string, unknown>;
}

export interface FinKpis {
  refunds: Partial<Record<RefundStatus, { count: number; amountMinor: number }>>;
  holds:   Partial<Record<HoldStatus,   { count: number; amountMinor: number }>>;
  ledger30d: Array<{ account: string; direction: 'credit' | 'debit'; amountMinor: number }>;
}
export interface FinInsight { id: string; severity: 'success' | 'info' | 'warn' | 'critical'; title: string }
export interface FinRiskScore {
  score: number; band: 'normal' | 'elevated' | 'high' | 'critical';
  model: string; factors: Record<string, number>;
}
export interface FinOverview {
  kpis: FinKpis; controls: FinControl[]; recentRefunds: FinRefund[]; holds: FinHold[];
  insights: FinInsight[]; riskScore: FinRiskScore; computedAt: string;
}
export interface FinList<T> { items: T[]; total: number; meta: { source: string; role: FinanceRole; page: number; pageSize: number } }

File: apps/ml-python/app/finance_admin.py
"""Domain 68 — Finance Admin risk scoring (deterministic, explainable)."""
from __future__ import annotations
from typing import Any, Dict

from fastapi import APIRouter
from pydantic import BaseModel, Field

router = APIRouter(prefix="/finance-admin", tags=["finance-admin"])


class ScoreIn(BaseModel):
    signals: Dict[str, Any] = Field(default_factory=dict)


@router.post("/score")
def score(body: ScoreIn):
    s = body.signals or {}
    refunds = s.get("refunds") or {}
    holds = s.get("holds") or {}
    pending = int((refunds.get("pending") or {}).get("count") or 0)
    processing = int((refunds.get("processing") or {}).get("count") or 0)
    failed = int((refunds.get("failed") or {}).get("count") or 0)
    active_holds = int((holds.get("active") or {}).get("count") or 0)

    score_v = 10 + min(40, (pending + processing) * 4) + min(30, failed * 6) + min(20, active_holds * 3)
    score_v = min(100, score_v)
    if score_v >= 80: band = "critical"
    elif score_v >= 60: band = "high"
    elif score_v >= 35: band = "elevated"
    else: band = "normal"
    return {
        "score": score_v, "band": band, "model": "deterministic-v1",
        "factors": {"pendingRefunds": pending + processing, "failedRefunds": failed, "activeHolds": active_holds},
    }
